game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java

```diff
@@ -235,8 +238,14 @@ public boolean createClientMessenger(final Component ui) {
               props,
               objectStreamFactory,
               new ClientLogin(this.ui, Injections.getInstance().getEngineVersion()));
-      // TODO: Project#20 replace no-op sender with a real sender.
-      clientNetworkBridge = ClientNetworkBridge.NO_OP_SENDER;
+
+      if (ClientSetting.useWebsocketNetwork.getValue().orElse(false)) {
+        final URI serverURI =
+            URI.create("ws://" + props.getHost() + ":" + ServerLauncher.RELAY_SERVER_PORT);
```

Comment:
This socket will be something that any player could theoretically open. Instead of hosting and opening a java server socket, they would open a websocket. Is it feasible to use SSL for any arbitrary TripleA game instance?

Reply:
Ah, I was assuming this way already the a "relay server" acting here, as in something managed by us, so I thought we could simply use letsencrypt here.
In any case TLS would be super nice to have, but obviously we can't get a trusted certificate for every user.
Using self-signed certificates would be an option (and therefore to simply trust self-signed certificates for this) to encrypt traffic, which would be better than sending everything in plain text, but the system would still be vulnerable to MITM attacks.
However I was wondering if we can establish trust in some other way. We might be able to setup the lobby as a mini-CA which hands out signed certificates we can check against the lobby public key we could create a some point. This doesn't solve the problem completely though, if we just started handing out certificates to anyone, MITM would still be possible, because an attacker could just use a real certificate to fake the connection. So at some point we need a control mechanism that checks the IP of a potential server when requesting a certificate for said IP.
Not sure how easy or difficult it is going to be to implement a system like that, but I think this the highest amount of security we can get without using actual CAs.
By using this approach we get full TLS security if our lobby server hands out a certificate to the right/legitimate client, if this step fails we have the same security like when using a self signed certificate.
Thoughts?

Reply:
The opening of a websocket will be a complete replacement of the current java sockets. We would want this to work for local networks that are behind a VPN, really preferably without any dependency whatsoever on the lobby.

Reply:
The relay server, to clarify, is a full and perhaps already complete replacement for bots. It just forwards any message it gets from one host to everyone that is connected to it. Any game host fires up a relay server and then 'self-connects' to it as would any client.

The lobby will dynamically create relay servers to transparently allow for anyone to host.